### PR TITLE
Added optional delays to starting controller managers. #22669

### PR DESCRIPTION
**Pull request by rrati**

### CODE REVIEW

**Comment by rrati** [comment]
controller-manager startup w/o no delays:
CPU:
cpu_no_delay


Memory:
memory_no_delay

controller-manager startup w/ 10 second delay:
CPU:
cpu_10s_delay

Memory:
memory_10s_delay

Interesting note is more memory is allocated when no delays are used.

**Comment by wojtek-t** [comment]
@rrati - so IIUC, the PR is effectively a no-op, right? To make the jittering effective you need to set the "--controller-manager-start-interval" flag, right?

**Comment by rrati** [comment]
@wojtek-t Correct.  The defaults are at 0 so it won't have an impact unless someone sets the delay.  I'm fine setting a default delay if we think that is the preferred path.  A 10s delay has a big impact.

**Comment by wojtek-t** [comment]
@rrati - I would prefer leaving the default 0 for now. We can change it later if needed.
Also - can you please check why tests are failing and fix them?

File: pkg/apis/componentconfig/types.go [hunk 1]
```diff
@@ -510,6 +510,10 @@ type KubeControllerManagerConfiguration struct {
 	LeaderElection LeaderElectionConfiguration `json:"leaderElection"`
 	// vloumeConfiguration holds configuration for volume related features.
 	VolumeConfiguration VolumeConfiguration `json:"volumeConfiguration"`
+	// How long to wait between starting controller managers
+	ControllerManagerStartInterval unversioned.Duration `json:"controllerManagerStartInterval"`
+	// Jitter for starting controller managers
+	ControllerManagerStartJitter float64 `json:"controllerManagerStartJitter"`
```
Comment (wojtek-t) [inline pkg/apis/componentconfig/types.go]: This one doesn't seem to be used anywhere. Can we remove it?

File: cmd/kube-controller-manager/app/options/options.go [hunk 2]
```diff
@@ -146,5 +147,6 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.RootCAFile, "root-ca-file", s.RootCAFile, "If set, this root certificate authority will be included in service account's token secret. This must be a valid PEM-encoded CA bundle.")
 	fs.Float32Var(&s.KubeAPIQPS, "kube-api-qps", s.KubeAPIQPS, "QPS to use while talking with kubernetes apiserver")
 	fs.IntVar(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
+	fs.DurationVar(&s.ControllerManagerStartInterval.Duration, "controller-manager-start-interval", s.ControllerManagerStartInterval.Duration, "Interval between starting controller managers.")
```
Comment (lavalamp) [inline cmd/kube-controller-manager/app/options/options.go]: The binary is controller-manager, the individual control loops are just "controllers".

File: pkg/apis/componentconfig/types.go (hunk 1 above)
Comment (lavalamp) [inline pkg/apis/componentconfig/types.go]: +1

**Comment by lavalamp** [comment]
We might want to think about controller startup order if we do this. Does the current ordering make sense?

**Comment by wojtek-t** [comment]
>
We might want to think about controller startup order if we do this. Does the current ordering make sense?

hmm - how changing the order can improve things?

File: cmd/kube-controller-manager/app/options/options.go (hunk 2 above)
Comment (rrati) [inline cmd/kube-controller-manager/app/options/options.go]: Changed

File: pkg/apis/componentconfig/types.go (hunk 1 above)
Comment (rrati) [inline pkg/apis/componentconfig/types.go]: Removed

**Comment by rrati** [comment]
@wojtek-t I'm not familiar with the cause for the GCE Node e2e build failure.  Is that an infra flake or can I repro locally?

**Comment by lavalamp** [comment]
I'm not sure, that's why I'm asking. If some important controller doesn't
start for a few minutes, is that an issue?

On Mon, Apr 4, 2016 at 5:52 PM, Wojciech Tyczynski <notifications@github.com

**Comment by wojtek-t** [comment]
@lavalamp - I think we don't have any strong dependencies; so that cluster may be not fully functional (i.e. for example replication controller is not creating pods), but I'm not aware of any long-standing issues it may cause

**Comment by wojtek-t** [comment]
@k8s-bot test this please github issue: #23878

**Comment by wojtek-t** [comment]
LGTM

**Comment by rrati** [comment]
Recent failures look like infra setup issues

**Comment by timothysc** [comment]
@rrati rebase and re-push, seems to be the standard way to bypass.

**Comment by timothysc** [comment]
re-applying lgtm

**Pull request merged**